File: 1025.cpp
#include <iostream>
#include <map>
#include <string>
#include <algorithm>
using namespace std;
map<string,int> lrank;
class Testee
{
public:
	string id;
	int local, score;
	Testee() {}
	Testee(string i,int l,int s)
	{ id = i; local = l; score = s; }
};
bool operator <(Testee A, Testee B)
{
	if (A.score == B.score)
		return A.id < B.id;
	return B.score < A.score;
}
Testee local[105][305], all[30005];
int lsum[105];
int main()
{
	cin.sync_with_stdio(0);
	string stuid,lastid = "0";
	int i,j,n,m,score,tot = 0,lasts = -1000,lastrk = -1,t;
	cin >> n;
	for (i = 1;i <= n;++i)
	{
		cin >> lsum[i];
		for (j = 1;j <= lsum[i];++j)
		{
			cin >> stuid >> score;
			local[i][j] = Testee(stuid,i,score);
			all[++tot] = Testee(stuid,i,score);
		}
		sort(local[i] + 1,local[i] + lsum[i] + 1);
		for (j = 1;j <= lsum[i];++j)
		{
			if (local[i][j].score == lasts)
				lrank[local[i][j].id] = lrank[lastid];
			else
				lrank[local[i][j].id] = j;
			lasts = local[i][j].score;
			lastid = local[i][j].id;
		}
	}
	sort(all + 1, all + tot + 1);
	cout << tot << endl;
	lasts = -1000;
	for (i = 1;i <= tot;++i)
	{
		cout << all[i].id << " ";
		cout << (t = (all[i].score == lasts ? lastrk : i)) << " ";
		cout << all[i].local << " ";
		cout << lrank[all[i].id] << endl;
		lasts = all[i].score; lastrk = t;
	}
	return 0;
}

File: 1025.rb
# TLE
class Testee
    attr_accessor :sid, :score, :local
    def initialize(sid, score, local)
        @sid = sid; @score = score; @local = local
    end
end
n = gets.chomp.to_i
local = Array.new(n) { Array.new }
all = []
lrank = {}
n.times { |i|
    m = gets.chomp.to_i
    m.times do
        sid, score = gets.chomp.split(" ")
        score = score.to_i
        all << Testee.new(sid, score, i + 1)
        local[i] << Testee.new(sid, score, i + 1)
    end
}
local.each_index { |i|
    local[i].sort! { |a, b| b.score <=> a.score }
    last = -1000; lastid = nil
    local[i].each_with_index { |j, index|
        lrank[j.sid] = (j.score == last ? lrank[lastid] : index + 1) 
        last = j.score; lastid = j.sid
    }
}
all.sort! { |a,b| a.score == b.score ? a.sid <=> b.sid : b.score <=> a.score }
puts all.length
last = -1000; lastrank = 0
all.each_index { |i|
    x = all[i]
    print x.sid , " "
    print t = (x.score == last ? lastrank : i + 1) , " "
    print x.local, " "
    print lrank[x.sid], "\n"
    last = x.score; lastrank = t
}

